cdap-tms/src/main/java/io/cdap/cdap/messaging/store/leveldb/LevelDBMessageTable.java

```diff
@@ -64,134 +74,102 @@ byte getType() {
+      Map<String, byte[]> columns = decodeValue(value);
+      messageTableKey.setFromRowKey(key);
+      return tableEntry.set(messageTableKey, columns.get(TX_COL), columns.get(PAYLOAD_COL));
     };
+
+    return new PartitionedDBScanIterator<>(partitions.iterator(), scanRequest.getStartRow(), scanRequest.getStopRow(),
+                                           decodeFunction);
   }
 
   @Override
   protected void persist(Iterator<RawMessageTableEntry> entries) throws IOException {
-    try (WriteBatch writeBatch = levelDB.createWriteBatch()) {
-      while (entries.hasNext()) {
-        RawMessageTableEntry entry = entries.next();
-        byte[] rowKey = entry.getKey();
-        // LevelDB doesn't make copies, and since we reuse RawMessageTableEntry object, we need to create copies.
-        writeBatch.put(Arrays.copyOf(rowKey, rowKey.length), encodeValue(entry.getTxPtr(), entry.getPayload()));
-      }
-      levelDB.write(writeBatch, WRITE_OPTIONS);
-    } catch (DBException ex) {
-      throw new IOException(ex);
+    if (!entries.hasNext()) {
+      return;
     }
-  }
 
-  @Override
-  protected void rollback(byte[] startKey, byte[] stopKey, byte[] txWritePtr) throws IOException {
-    WriteBatch writeBatch = levelDB.createWriteBatch();
-    try (CloseableIterator<Map.Entry<byte[], byte[]>> rowIterator = new DBScanIterator(levelDB, startKey, stopKey)) {
-      while (rowIterator.hasNext()) {
-        Map.Entry<byte[], byte[]> rowValue = rowIterator.next();
-        byte[] value = rowValue.getValue();
-        Map<String, byte[]> columns = decodeValue(value);
-        writeBatch.put(rowValue.getKey(), encodeValue(txWritePtr, columns.get(PAYLOAD_COL)));
+    // entries are sorted by publish time. accumulate all entries for a partition into a batch and
+    // write the batch when the next entry is outside of the current partition
+    LevelDBPartition partition = null;
+    WriteBatch writeBatch = null;
+    while (entries.hasNext()) {
+      RawMessageTableEntry entry = entries.next();
+      byte[] rowKey = entry.getKey().getRowKey();
+      long publishTime = entry.getKey().getPublishTimestamp();
+
+      // check if this entry belongs in a different partition. If so, write the current batch.
+      if (partition == null || publishTime < partition.getStartTime() || publishTime >= partition.getEndTime()) {
+        if (partition != null) {
+          try {
+            partition.getLevelDB().write(writeBatch, WRITE_OPTIONS);
+          } finally {
+            writeBatch.close();
+          }
+        }
+        partition = partitionManager.getOrCreatePartition(publishTime);
+        writeBatch = partition.getLevelDB().createWriteBatch();
       }
+
+      // LevelDB doesn't make copies, and since we reuse RawMessageTableEntry object, we need to create copies.
+      writeBatch.put(Arrays.copyOf(rowKey, rowKey.length), encodeValue(entry.getTxPtr(), entry.getPayload()));
     }
 
     try {
```

Comment:
Add a `if (partition != null)` check to safeguard.

Reply:
this would generate a warning because of the entires.hasNext() check at the start. I'll remove that check though, since a check here would be less error-prone moving forward.